File: java/src/test/java/bootcamp/ContractTests.java
package bootcamp;

import net.corda.core.identity.CordaX500Name;
import net.corda.testing.contracts.DummyState;
import net.corda.testing.core.DummyCommandData;
import net.corda.testing.core.TestIdentity;
import net.corda.testing.node.MockServices;
import org.junit.Test;

import static net.corda.testing.node.NodeTestUtils.ledger;

public class ContractTests {
    private final TestIdentity alice = new TestIdentity(new CordaX500Name("Alice", "", "GB"));
    private final TestIdentity bob = new TestIdentity(new CordaX500Name("Bob", "", "GB"));
    private MockServices ledgerServices = new MockServices(new TestIdentity(new CordaX500Name("TestId", "", "GB")));
    private TokenState tokenState = new TokenState(alice.getParty(), bob.getParty(), 1);

    @Test
    public void tokenContractRequiresZeroInputsInTheTransaction() {
        ledger(ledgerServices, ledger -> {
            ledger.transaction(tx -> {
                // Has an input, will fail.
                tx.input(TokenContract.ID, tokenState);
                tx.output(TokenContract.ID, tokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.fails();
                return null;
            });

            ledger.transaction(tx -> {
                // Has no input, will verify.
                tx.output(TokenContract.ID, tokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.verifies();
                return null;
            });

            return null;
        });
    }

    @Test
    public void tokenContractRequiresOneOutputInTheTransaction() {
        ledger(ledgerServices, ledger -> {
            ledger.transaction(tx -> {
                // Has two outputs, will fail.
                tx.output(TokenContract.ID, tokenState);
                tx.output(TokenContract.ID, tokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.fails();
                return null;
            });

            ledger.transaction(tx -> {
                // Has one input, will verify.
                tx.output(TokenContract.ID, tokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.verifies();
                return null;
            });

            return null;
        });
    }

    @Test
    public void tokenContractRequiresOneCommandInTheTransaction() {
        ledger(ledgerServices, ledger -> {
            ledger.transaction(tx -> {
                tx.output(TokenContract.ID, tokenState);
                // Has two commands, will fail.
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.fails();
                return null;
            });

            ledger.transaction(tx -> {
                tx.output(TokenContract.ID, tokenState);
                // Has one command, will verify.
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.verifies();
                return null;
            });

            return null;
        });
    }

    @Test
    public void tokenContractRequiresTheTransactionsOutputToBeATokenState() {
        ledger(ledgerServices, ledger -> {
            ledger.transaction(tx -> {
                // Has wrong output type, will fail.
                tx.output(TokenContract.ID, new DummyState());
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.fails();
                return null;
            });

            ledger.transaction(tx -> {
                // Has correct output type, will verify.
                tx.output(TokenContract.ID, tokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.verifies();
                return null;
            });

            return null;
        });
    }

    @Test
    public void tokenContractRequiresTheTransactionsOutputToHaveAPositiveAmount() {
        TokenState zeroTokenState = new TokenState(alice.getParty(), bob.getParty(), -1);
        TokenState negativeTokenState = new TokenState(alice.getParty(), bob.getParty(), -1);
        TokenState positiveTokenState = new TokenState(alice.getParty(), bob.getParty(), 2);

        ledger(ledgerServices, ledger -> {
            ledger.transaction(tx -> {
                // Has zero-amount TokenState, will fail.
                tx.output(TokenContract.ID, zeroTokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.fails();
                return null;
            });

            ledger.transaction(tx -> {
                // Has negative-amount TokenState, will fail.
                tx.output(TokenContract.ID, negativeTokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.fails();
                return null;
            });

            ledger.transaction(tx -> {
                // Has positive-amount TokenState, will verify.
                tx.output(TokenContract.ID, tokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.verifies();
                return null;
            });

            ledger.transaction(tx -> {
                // Also has positive-amount TokenState, will verify.
                tx.output(TokenContract.ID, positiveTokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.verifies();
                return null;
            });

            return null;
        });
    }

    @Test
    public void tokenContractRequiresTheTransactionsCommandToBeAnIssueCommand() {
        ledger(ledgerServices, ledger -> {
            ledger.transaction(tx -> {
                // Has wrong command type, will fail.
                tx.output(TokenContract.ID, tokenState);
                tx.command(alice.getPublicKey(), DummyCommandData.INSTANCE);
                tx.fails();
                return null;
            });

            ledger.transaction(tx -> {
                // Has correct command type, will verify.
                tx.output(TokenContract.ID, tokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.verifies();
                return null;
            });

            return null;
        });
    }

    @Test
    public void tokenContractRequiresTheIssuerToBeARequiredSignerInTheTransaction() {
        TokenState tokenStateWhereBobIsIssuer = new TokenState(bob.getParty(), alice.getParty(), 1);

        ledger(ledgerServices, ledger -> {
            ledger.transaction(tx -> {
                // Issuer is not a required signer, will fail.
                tx.output(TokenContract.ID, tokenState);
                tx.command(bob.getPublicKey(), DummyCommandData.INSTANCE);
                tx.fails();
                return null;
            });

            ledger.transaction(tx -> {
                // Issuer is also not a required signer, will fail.
                tx.output(TokenContract.ID, tokenStateWhereBobIsIssuer);
                tx.command(alice.getPublicKey(), DummyCommandData.INSTANCE);
                tx.fails();
                return null;
            });

            ledger.transaction(tx -> {
                // Issuer is a required signer, will verify.
                tx.output(TokenContract.ID, tokenState);
                tx.command(alice.getPublicKey(), new TokenContract.Issue());
                tx.verifies();
                return null;
            });

            ledger.transaction(tx -> {
                // Issuer is also a required signer, will verify.
                tx.output(TokenContract.ID, tokenStateWhereBobIsIssuer);
                tx.command(bob.getPublicKey(), new TokenContract.Issue());
                tx.verifies();
                return null;
            });

            return null;
        });
    }
}

File: src/test/kotlin/bootcamp/ContractTests.kt
package bootcamp

import net.corda.testing.*
import net.corda.testing.contracts.DummyState
import org.junit.After
import org.junit.Before
import org.junit.Test

class ContractTests {
    private val tokenState = TokenState(ALICE, BOB, 1)

    @Before
    fun setup() {
        setCordappPackages("bootcamp")
    }

    @After
    fun tearDown() {
        unsetCordappPackages()
    }

    @Test
    fun `TokenContract imposes constraint that there are zero inputs`() {
        ledger {
            transaction {
                // Has an input, will fail.
                input(TokenContract.ID) { tokenState }
                output(TokenContract.ID) { tokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                fails()
            }
            transaction {
                // Has no input, will verify.
                output(TokenContract.ID) { tokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                verifies()
            }
        }
    }

    @Test
    fun `TokenContract imposes constraint that there is one output`() {
        ledger {
            transaction {
                // Has two outputs, will fail.
                output(TokenContract.ID) { tokenState }
                output(TokenContract.ID) { tokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                fails()
            }
            transaction {
                // Has one input, will verify.
                output(TokenContract.ID) { tokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                verifies()
            }
        }
    }

    @Test
    fun `TokenContract imposes constraint that there is one command`() {
        ledger {
            transaction {
                output(TokenContract.ID) { tokenState }
                // Has two commands, will fail.
                command(ALICE_PUBKEY) { TokenContract.Issue }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                fails()
            }
            transaction {
                output(TokenContract.ID) { tokenState }
                // Has one command, will verify.
                command(ALICE_PUBKEY) { TokenContract.Issue }
                verifies()
            }
        }
    }

    @Test
    fun `TokenContract imposes constraint that the output is a TokenState`() {
        ledger {
            transaction {
                // Has wrong output type, will fail.
                output(TokenContract.ID) { DummyState() }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                fails()
            }
            transaction {
                // Has correct output type, will verify.
                output(TokenContract.ID) { tokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                verifies()
            }
        }
    }

    @Test
    fun `TokenContract imposes constraint that the output TokenState has a positive amount`() {
        val zeroTokenState = TokenState(ALICE, BOB, -1)
        val negativeTokenState = TokenState(ALICE, BOB, -1)
        val positiveTokenState = TokenState(ALICE, BOB, 2)

        ledger {
            transaction {
                // Has zero-amount TokenState, will fail.
                output(TokenContract.ID) { zeroTokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                fails()
            }
            transaction {
                // Has negative-amount TokenState, will fail.
                output(TokenContract.ID) { negativeTokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                fails()
            }
            transaction {
                // Has positive-amount TokenState, will verify.
                output(TokenContract.ID) { tokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                verifies()
            }
            transaction {
                // Also has positive-amount TokenState, will verify.
                output(TokenContract.ID) { positiveTokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                verifies()
            }
        }
    }

    @Test
    fun `TokenContract imposes constraint that the command is an Issue command`() {
        ledger {
            transaction {
                // Has wrong command type, will fail.
                output(TokenContract.ID) { tokenState }
                command(ALICE_PUBKEY) { DummyCommandData }
                fails()
            }
            transaction {
                // Has correct command type, will verify.
                output(TokenContract.ID) { tokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                verifies()
            }
        }
    }

    @Test
    fun `TokenContract imposes constraint that the issuer is a required signer`() {
        val tokenStateWhereBobIsIssuer = TokenState(BOB, ALICE, 1)

        ledger {
            transaction {
                // Issuer is not a required signer, will fail.
                output(TokenContract.ID) { tokenState }
                command(BOB_PUBKEY) { DummyCommandData }
                fails()
            }
            transaction {
                // Issuer is also not a required signer, will fail.
                output(TokenContract.ID) { tokenStateWhereBobIsIssuer }
                command(ALICE_PUBKEY) { DummyCommandData }
                fails()
            }
            transaction {
                // Issuer is a required signer, will verify.
                output(TokenContract.ID) { tokenState }
                command(ALICE_PUBKEY) { TokenContract.Issue }
                verifies()
            }
            transaction {
                // Issuer is also a required signer, will verify.
                output(TokenContract.ID) { tokenStateWhereBobIsIssuer }
                command(BOB_PUBKEY) { TokenContract.Issue }
                verifies()
            }
        }
    }
}